service/controller/resource/certificates/resource.go

```diff
@@ -21,59 +23,82 @@ type Config struct {
 	Logger    micrologger.Logger
 }
 
-type Resource struct {
-	k8sClient k8sclient.Interface
-	logger    micrologger.Logger
+// secretCopier provides a `ToCR` method which copies data from the source
+// cluster secret CR
+type secretCopier struct {
+	clientFunc func(string) generic.Interface
 }
 
-func New(config Config) (*Resource, error) {
+func New(config Config) (*generic.Resource, error) {
 	if config.K8sClient == nil {
 		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
 	}
 	if config.Logger == nil {
 		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
 	}
 
-	r := &Resource{
-		k8sClient: config.K8sClient,
-		logger:    config.Logger,
+	clientFunc := func(namespace string) generic.Interface {
+		c := config.K8sClient.K8sClient().CoreV1().Secrets(namespace)
+		return wrappedClient{client: c}
 	}
 
-	return r, nil
-}
+	sc := secretCopier{clientFunc: clientFunc}
+
+	c := generic.Config{
+		ClientFunc:     clientFunc,
+		Logger:         config.Logger,
+		Name:           Name,
+		ToCR:           sc.ToCR,
```

Comment:
Nice!

Reply:
I just had a thought that `ToCR` could also be a closure defined inside `New` so there wouldn't be a need for the struct, not sure if it's worth changing though.

Reply:
True, first I wondered what was this struct for. Maybe its simpler with the closure, I have no strong opinion on this.